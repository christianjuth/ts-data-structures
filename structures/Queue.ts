/**
 * A data structure to store a collection of objects.
 *
 * @summary queue
 * @author Christian Juth
 *
 * Created at     : 2019-11-21 23:37:56 
 * Last modified  : 2019-11-22 00:38:39
 */


/**
 * A data structure to store a collection of objects.
 * 
 * @typeparam T - Type of value contained in the queue
 */
class Queue<T>{
  private front: Node<T> | null;
  private back: Node<T> | null;

  /**
   * Create a new queue.
   */
  constructor() {
    this.front = null;
    this.back = null;
  }

  /**
   * Add value to the queue.
   *
   * @param value - value to enqueue
   */
  enqueue(value: T): void {
    let node = new Node({ value });
    if(this.front === null) {
      this.back = node;
    } else {
      this.front.prev = node;
    }
    this.front = node;
  }

  /**
   * First node in queue, if it exists
   *
   * @return first value in list, if it exists
   */
  first(): T | null{
    if(this.back === null) return null;
    return this.back.value;
  }

  /**
   * Remove value from the queue.
   *
   * @return value that was dequeued
   */
  dequeue(): T | null{
    if(this.back === null) return null;
    const output = this.back.value;
    this.back = this.back.prev;
    return output;
  }

  /**
   * Check if the queue is empty
   *
   * @return true if the queue is empty
   */
  isEmpty(): boolean {
    return this.front === null;
  }

  /**
   * String representation of queue
   *
   * @return string representing queue
   */
  toString(): String {
    let output = '';
    for(let n = this.back; n; n = n.prev) {
      output += (n.value + (n.prev ? ', ' : ''));
    }
    return `(${output})`;
  }
  
  /**
   * Print string representation of queue to the console
   */
  traverse(): void {
    console.log(this.toString());
  }
}


interface Node<T> {
  value: T;
  prev: Node<T> | null;
}

class Node<T> implements Node<T>{
  constructor({ value, prev }: { value: T, prev?: Node<T> | null }) {
    this.value = value;
    // turn undefined into null
    this.prev = prev || null;
  }
}


export default Queue;