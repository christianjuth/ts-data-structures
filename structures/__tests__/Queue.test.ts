import Queue from '../Queue';


describe("Queue", () => {

  test("enqueue", () => {
    let queue = new Queue<number>();
    queue.enqueue(1);
  });

  test("first", () => {
    let queue = new Queue<String>();
    queue.enqueue('one');
    expect(queue.first()).toBe('one');
  });

  test("dequeue", () => {
    let queue = new Queue<String>();
    queue.enqueue('one');
    queue.enqueue('two');
    expect(queue.dequeue()).toBe('one');
    expect(queue.dequeue()).toBe('two');
  });

  test("isEmpty", () => {
    let queue = new Queue<number>();
    expect(queue.isEmpty()).toBe(true);
    queue.enqueue(0);
    expect(queue.isEmpty()).toBe(false);
  });

  test("toString", () => {
    let queue = new Queue<String>();
    queue.enqueue('one');
    queue.enqueue('two');
    queue.enqueue('three');
    queue.enqueue('four');
    expect(queue.toString()).toMatchSnapshot();
  });

});