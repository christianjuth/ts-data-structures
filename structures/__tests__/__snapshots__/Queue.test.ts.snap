// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Queue toString 1`] = `"(one, two, three, four)"`;
